### DISCUSSION TITLE
[question] What are the ONCE and LOWCODE segments in general and what are it for the c16 target?

### DISCUSSION THREAD

**Discussion by Fabrizio-Caruso** [cc65/cc65#2140] [Q&A]
What are the ONCE and LOWCODE segments in general and what are they for the C16 target?

Can I use LOWCODE to some code in a different area other than the usual contiguous 11k on the C16 memory?
How can the loader know that some of the code has to be moved elsewhere?

**Comment by mrdudz** [comment]
> What are the ONCE and LOWCODE segments in general and what are they for the C16 target?

Generally ONCE is reserved for init code, which is run once at startup and then can be discarded (overwritten). Typically it is located at the end of the binary, and will become part of BSS after the init code was run.

LOWCODE is reserved for code that must be "banked in" at all times. Eg on the C64 code that calls into the BASIC ROM must go into LOWCODE, else that code might end up "under" the BASIC ROM itself (which can't work). For many targets this is simply located at the beginning at the binary, as that will typically end up in a section of memory that is not "banked".

> Can I use LOWCODE to some code in a different area other than the usual contiguous 11k on the C16 memory?

Not sure what you mean. It can be located anywhere - if it is "banked in" at all times.

> How can the loader know that some of the code has to be moved elsewhere?

This "knowledge" is in the crt0 and linker config. For the CBM targets anyway :) If a segment needs to be copied to a different location, the crt0 has to do it.

**Comment by Fabrizio-Caruso** [comment]
@mrdudz So, there is no automatic/magic code that will instruct the loader to move the code to LOWCODE.
I mean, I have to write a special LOADER and modify crt0.s in order to use LOWCODE in a target like the C16.

Having a crt with:
```
...
    TAPEBUF:    file = "", define = yes, start = $0333, size = $0100;
...
}
SEGMENTS {
...
    LOWCODE:  load = TAPEBUF,       type = ro,  optional = yes;
```
and setting some pragma or link-time instructions to have some code in LOWCODE, will not create a binary that self-loads some portions into LOWCODE.

**Comment by mrdudz** [comment]
exactly, in the default config there is no need to copy it around, so there is no code for it

**Comment by Fabrizio-Caruso** [comment]
@mrdudz my current issue is that I would like to gain a few bytes by placing some code elsewhere. I don't know how to do it. I would have to rewrite the crt0.s in order to copy some code in the tape buffer or elsewhere.
Additionally I would like to have some selected variables use some of the free (non-contiguous) memory.
There is no way to do this with CC65 without polluting the source code with pragmas or other directives.
In my case I won't to keep the same code for all my targets (most of which are not MOS 6502 based).
With the cfg linker file I can only place entire segment in specific memory areas but for example I cannot split BBS and put half of it into the tape buffer and the other half elsewhere by selecting which variables have to go to a specific memory area.

**Comment by mrdudz** [comment]
There is no other way if you want to rearrange memory like this
